### PR TITLE
Adds Stories class to list stories in different view formats

### LINKED ISSUE #14: Template Tags - Provision of Public API [docs]
Provide public API that theme developers can use to easily pull web stories into different parts of the website such as header, footer, or any other area.

The deliverable will be documentation that we can give to theme developers to help them integrate Web Stories into their theme easily

### PR DESCRIPTION
**Pull request by deepaklalwani97**
## Summary

- Adds a Stories class that can be used by users to list stories in different view formats based on the attributes provided.

## Relevant Technical Choices

- Added Stories class which uses the Renderer classes to display stories based on given attributes.
- Added a Generic Renderer class which handles the circles, grid, list, and carousel story view types.

## Testing Instructions

Displaying stories with default arguments.
```php
$stories = new Stories();
echo $stories->render();
```

Displaying Stories with user-provided arguments.
```php
$story_attributes = [
	'view_type'         => 'carousel',
	'show_story_poster' => false,
	'show_title'        => true,
];

$query_arguments = [
	'posts_per_page' => 10,
	'order'          => 'ASC',
	'orderby'        => 'title',
];
$stories = new Stories( $story_attributes, $query_arguments );
echo $stories->render();
```

Fixes #14

### CODE REVIEW

File: includes/Interfaces/Renderer.php [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+namespace Google\Web_Stories\Interfaces;
+
+// Prevent direct access.
+defined( 'ABSPATH' ) || exit;
```
Comment (swissspidy) [inline includes/Interfaces/Renderer.php:31]: This is not something we currently do in the code base, FWIW. Don't think we need to start doing that.

File: includes/Stories_Renderer/Generic_Renderer.php [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+			do_action( 'web_stories_renderer_wrapper_end' );
+			?>
+		</div>
+		<?php
+	}
+
+	/**
+	 * Add amp-carousel for carousel view type.
+	 */
+	public function amp_carousel() {
```
Comment (pdclark) [inline includes/Stories_Renderer/Generic_Renderer.php:153]: Add `@return null` re: PHPCS error line 153

File: includes/Stories_Renderer/Generic_Renderer.php [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+				?>
+				</amp-carousel>
+				<?php
+		}
+	}
+
+	/**
+	 * Renders stories archive link if the 'show_view_all_link' attribute is set to true.
+	 */
+	public function maybe_render_archive_link() {
```
Comment (pdclark) [inline includes/Stories_Renderer/Generic_Renderer.php:197]: Add `@return null` re: PHPCS Line 197

File: includes/Stories_Renderer/Generic_Renderer.php [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+
+		if ( ( ! empty( $this->attributes['show_view_all_link'] ) ) && ( true === $this->attributes['show_view_all_link'] ) ) :
+			$web_stories_archive = get_post_type_archive_link( Story_Post_Type::POST_TYPE_SLUG );
+			?>
+			<div class="web-stories__archive-link">
+				<a href="<?php echo( esc_url_raw( $web_stories_archive ) ); ?>">
```
Comment (pdclark) [inline includes/Stories_Renderer/Generic_Renderer.php:203]: Add PHPCS Ignore?

File: includes/Stories_Renderer/Generic_Renderer.php [hunk 5]
```diff
@@ -0,0 +1,387 @@
...
+		</div>
+		<?php
+	}
+
+	/**
+	 * Renders a story with story's poster image.
+	 *
+	 * @param array $story_data Story item data. Contains information like url, height, width, etc of the story.
+	 */
+	public function render_story_with_poster( array $story_data ) {
```
Comment (pdclark) [inline includes/Stories_Renderer/Generic_Renderer.php:294]: Add `@return null` re: PHPCS line 294

File: includes/Stories_Renderer/Generic_Renderer.php [hunk 6]
```diff
@@ -0,0 +1,387 @@
...
+			<?php
+		}
+	}
+
+	/**
+	 * Renders a story with amp-story-player.
+	 *
+	 * @param array $story_data Story attributes. Contains information like url, height, width, etc of the story.
+	 */
+	public function render_story_with_story_player( array $story_data ) {
```
Comment (pdclark) [inline includes/Stories_Renderer/Generic_Renderer.php:328]: Add @return null re: PHPCS line 328

File: includes/Stories_Renderer/Generic_Renderer.php [hunk 7]
```diff
@@ -0,0 +1,387 @@
...
+			do_action( 'web_stories_renderer_story_content_overlay', $story_data );
+		}
+	}
+
+	/**
+	 * Renders the content overlay markup.
+	 *
+	 * @param array $story_data Story item data. Contains information like url, height, width, etc of the story.
+	 */
+	public function get_content_overlay( array $story_data ) {
```
Comment (pdclark) [inline includes/Stories_Renderer/Generic_Renderer.php:351]: Add @return null re: PHPCS line 351

**Review by pdclark — CHANGES_REQUESTED** [review]

**Review by pdclark — APPROVED** [review]

File: includes/Stories_Renderer/Generic_Renderer.php [hunk 8]
```diff
@@ -0,0 +1,390 @@
...
+		add_action( 'web_stories_renderer_container_start', [ $this, 'amp_carousel' ] );
+		add_action( 'web_stories_renderer_container_end', [ $this, 'amp_carousel' ] );
+		add_action( 'web_stories_renderer_content', [ $this, 'render_story' ] );
+		add_action( 'web_stories_renderer_wrapper_end', [ $this, 'maybe_render_archive_link' ] );
+		add_action( 'web_stories_renderer_story_content_overlay', [ $this, 'get_content_overlay' ] );
+		add_action( 'web_stories_renderer_single_story_content', [ $this, 'render_story_with_poster' ] );
+		add_action( 'web_stories_renderer_single_story_content', [ $this, 'render_story_with_story_player' ] );
+		add_action( 'web_stories_renderer_container_classes', [ $this, 'container_classes' ] );
+		add_action( 'web_stories_renderer_container_style', [ $this, 'container_styles' ] );
+		add_action( 'web_stories_renderer_single_story_classes', [ $this, 'single_story_classes' ] );
```
Comment (swissspidy) [inline includes/Stories_Renderer/Generic_Renderer.php:52]: Interesting approach to templating but seems very hard to follow, test, and maintain.

I'd split this class up into multiple ones, e.g. Carousel, List, Grid, etc., then the `Stories` class can instantiate the right one

File: includes/Interfaces/Renderer.php [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+
+interface Renderer {
+
+	/**
+	 * Initial actions to setup the renderer like,
+	 * adding hooks and setting up states.
+	 *
+	 * @return void
+	 */
+	public function setup();
```
Comment (swissspidy) [inline includes/Interfaces/Renderer.php:38]: Nit: we usually use `init()` everywhere else

File: includes/Stories_Renderer/Generic_Renderer.php (hunk 8 above)
Comment (ankitrox) [reply includes/Stories_Renderer/Generic_Renderer.php:52]: @swissspidy We have almost everything similar in case of all other view types but Carousel, so we have split classes into `Generic_Renderer` and `Carousel_Renderer`, where `Carousel_renderer` would be responsible for rendering Carousel stories, but others would be responsible for List, Grid etc. 

We can, of course, create other classes unless there is a significant difference between them.

Let me know your opinion.

Comment (swissspidy) [reply includes/Stories_Renderer/Generic_Renderer.php:52]: I don't see a `Carousel_Renderer` class here though?

File: includes/Stories_Renderer/Generic_Renderer.php [hunk 10]
```diff
@@ -0,0 +1,389 @@
...
```
Comment (swissspidy) [inline includes/Stories_Renderer/Generic_Renderer.php:52]: Re-iterating on my initial statement:

Templating by doing `add_action` and `current_action` checks everywhere is very hard to follow and maintain. For example, when reading the code it's hard to find out what the final carousel template looks like.

File: includes/Stories_Renderer/Renderer.php [hunk 11]
```diff
@@ -0,0 +1,368 @@
...
+	 * @return string
+	 */
+	protected function get_container_classes() {
+		$container_classes  = 'web-stories ';
+		$container_classes .= ( ! empty( $this->attributes['view_type'] ) ) ? " is-view-type-{$this->attributes['view_type']}" : ' is-view-type-grid';
+		$container_classes .= ( ! empty( $this->attributes['align'] ) ) ? " align{$this->attributes['align']}" : ' alignnone';
+
+		$classes = ! empty( $this->attributes['classes'] ) ? $this->attributes['classes'] : '';
+
+		return trim( $classes . ' ' . $container_classes );
```
Comment (swissspidy) [inline includes/Stories_Renderer/Renderer.php:212]: Seems easier to do something like this:

```php
$container_classes = [];
$container_classes[] = 'web-stories';
$container_classes[] = ...

// ....

return implode(...)
```

(Aside: `web-stories` is a very generic class name. I'd use something more specific)

File: includes/Stories_Renderer/Renderer.php [hunk 12]
```diff
@@ -0,0 +1,368 @@
...
+		return apply_filters( 'web_stories_renderer_single_story_classes', $single_story_classes );
+	}
+
+	/**
+	 * Gets the container style attributes.
+	 *
+	 * @return string
+	 */
+	protected function get_container_styles() {
+		$container_style = ( true === $this->is_view_type( 'grid' ) ) ? "grid-template-columns:repeat({$this->attributes['number_of_columns']}, 1fr);" : '';
```
Comment (swissspidy) [inline includes/Stories_Renderer/Renderer.php:239]: I'd use `sprintf` for better readability.

File: includes/Stories_Renderer/Renderer.php [hunk 13]
```diff
@@ -0,0 +1,368 @@
...
+				echo( esc_html( $story_data['title'] ) );
+				?>
+			</div>
+			<?php endif; ?>
+			<div class="story-content-overlay__author-date">
+				<?php if ( ! empty( $story_data['author'] ) ) : ?>
+				<div>
+					<?php
+					esc_html_e( 'By', 'web-stories' );
+					echo( esc_html( ' ' . $story_data['author'] ) );
```
Comment (swissspidy) [inline includes/Stories_Renderer/Renderer.php:352]: String concatenation is a bad practice in internationalization.

Better:

```php
/* translators: %s: author name. */
echo esc_html( sprintf( __( 'By %s', 'web-stories' ), $story_data['author'] ) );
```

same for the publish date below

File: includes/Stories_Renderer/Renderer.php [hunk 14]
```diff
@@ -0,0 +1,368 @@
...
+					?>
+					</div>
+				<?php endif; ?>
+				<?php if ( ! empty( $story_data['date'] ) ) : ?>
+				<time class="story-content-overlay__date">
+					<?php
+					esc_html_e( 'On', 'web-stories' );
+					echo( esc_html( ' ' . $story_data['date'] ) );
```
Comment (swissspidy) [inline includes/Stories_Renderer/Renderer.php:360]:
```php
/* translators: %s: publish date. */
echo esc_html( sprintf( __( 'On %s', 'web-stories' ), $story_data['date'] ) );
```

File: includes/Stories_Renderer/Renderer.php [hunk 15]
```diff
@@ -0,0 +1,368 @@
...
+	 */
+	protected function maybe_render_archive_link() {
+		if ( ( ! empty( $this->attributes['show_stories_archive_link'] ) ) && ( true === $this->attributes['show_stories_archive_link'] ) ) :
+			$web_stories_archive = get_post_type_archive_link( Story_Post_Type::POST_TYPE_SLUG );
+			$web_stories_archive = ( is_string( $web_stories_archive ) ? $web_stories_archive : '' );
+			?>
+			<div class="web-stories__archive-link">
+				<a href="<?php echo( esc_url_raw( $web_stories_archive ) ); ?>">
+					<?php echo( esc_html( $this->attributes['stories_archive_label'] ) ); ?>
+				</a>
```
Comment (swissspidy) [inline includes/Stories_Renderer/Renderer.php:194]: If `$web_stories_archive` is an empty string (`''`), we'll end up with `<a href="">`.

Let's not print any link if there is none.

File: includes/Stories_Renderer/Generic_Renderer.php (hunk 8 above)
Comment (swissspidy) [reply includes/Stories_Renderer/Generic_Renderer.php]: @deepaklalwani97 I see you have changed things around now 👍

Comment (deepaklalwani97) [reply includes/Stories_Renderer/Generic_Renderer.php]: Yes, we have restructured the code for the renderer classes.

File: includes/Stories_Renderer/Renderer.php (hunk 11 above)
Comment (deepaklalwani97) [reply includes/Stories_Renderer/Renderer.php:212]: Yes, we will change the class name to be more specific.

File: includes/Stories_Renderer/Renderer.php [hunk 16]
```diff
@@ -0,0 +1,386 @@
...
+	 * @param int    $story_id             Story's id for which the story attributes are requested.
+	 * @param string $single_story_classes Single story's classes.
+	 *
+	 * @return array Returns single story item data.
+	 */
+	protected function get_story_item_data( $story_id, $single_story_classes = '' ) {
+
+		$story_data = [];
+
+		if ( empty( $story_id ) ) {
```
Comment (dhaval-parekh) [inline includes/Stories_Renderer/Renderer.php:132]: Also, Make sure `$story_id` is a positive number.

```
if ( empty( $story_id ) || 0 >= intval( $story_id ) ) {
    return $story_data;
}
```

File: includes/Stories_Renderer/Renderer.php [hunk 17]
```diff
@@ -0,0 +1,386 @@
...
+		$author_id       = absint( get_post_field( 'post_author', $story_id ) );
+		$is_circles_view = $this->is_view_type( 'circles' );
+		$image_size      = $is_circles_view ? Media::POSTER_SQUARE_IMAGE_SIZE : Media::POSTER_PORTRAIT_IMAGE_SIZE;
+		$story_title     = ( ! empty( $this->attributes['show_title'] ) && ( true === $this->attributes['show_title'] ) ) ?
+			get_the_title( $story_id ) :
+			'';
+		$author_name     = ( ! $is_circles_view && ! empty( $this->attributes['show_author'] ) && ( true === $this->attributes['show_author'] ) ) ?
+			get_the_author_meta( 'display_name', $author_id ) :
+			'';
+		$story_date      = ( ! $is_circles_view && ! empty( $this->attributes['show_date'] ) && ( true === $this->attributes['show_date'] ) ) ?
```
Comment (dhaval-parekh) [inline includes/Stories_Renderer/Renderer.php:145]: Better to use the `if...else` statement for complex conditions for readability.

File: includes/Stories_Renderer/Renderer.php [hunk 18]
```diff
@@ -0,0 +1,386 @@
...
+		$web_stories_archive = get_post_type_archive_link( Story_Post_Type::POST_TYPE_SLUG );
+
+		if ( empty( $web_stories_archive ) || ! is_string( $web_stories_archive ) ) {
+			return;
+		}
+
+		?>
+		<div class="web-stories-list__archive-link">
+			<a href="<?php echo( esc_url_raw( $web_stories_archive ) ); ?>">
+				<?php echo( esc_html( $this->attributes['stories_archive_label'] ) ); ?>
```
Comment (dhaval-parekh) [inline includes/Stories_Renderer/Renderer.php:202]: You might want to remove the round bracket from the `echo` statement.

File: includes/Stories_Renderer/Renderer.php [hunk 19]
```diff
@@ -0,0 +1,386 @@
...
+
+		$web_stories_archive = get_post_type_archive_link( Story_Post_Type::POST_TYPE_SLUG );
+
+		if ( empty( $web_stories_archive ) || ! is_string( $web_stories_archive ) ) {
+			return;
+		}
+
+		?>
+		<div class="web-stories-list__archive-link">
+			<a href="<?php echo( esc_url_raw( $web_stories_archive ) ); ?>">
```
Comment (dhaval-parekh) [inline includes/Stories_Renderer/Renderer.php:201]: Please use `esc_url()` for escaping URLs.

File: includes/Stories_Renderer/Renderer.php [hunk 20]
```diff
@@ -0,0 +1,386 @@
...
+	/**
+	 * Gets the single story container classes.
+	 *
+	 * @return string
+	 */
+	protected function get_single_story_classes() {
+
+		$single_story_classes   = [];
+		$single_story_classes[] = 'web-stories-list__story-wrapper';
+		$single_story_classes[] = ! $this->is_view_type( 'grid' ) ? 'has-poster' : '';
```
Comment (dhaval-parekh) [inline includes/Stories_Renderer/Renderer.php:234]: Not blocker: Use a round bracket around the condition.

```
$single_story_classes[] = ( ! $this->is_view_type( 'grid' ) ) ? 'has-poster' : '';
```

File: includes/Stories_Renderer/Renderer.php [hunk 21]
```diff
@@ -0,0 +1,386 @@
...
+		$single_story_classes[] = $this->is_view_type( 'grid' ) && ! empty( $this->attributes['show_story_poster'] && true === $this->attributes['show_story_poster'] ) ?
```
Comment (dhaval-parekh) [inline includes/Stories_Renderer/Renderer.php:235]: Better to use the `if...else` statement here.

Please update here and anywhere there is a longer condition statement or function is used within it.

File: includes/Stories_Renderer/Renderer.php [hunk 22]
```diff
@@ -0,0 +1,386 @@
...
+	 * Render story markup.
+	 *
+	 * @param int $story_id Story ID.
+	 *
+	 * @return void
+	 */
+	protected function render_single_story_content( $story_id ) {
+		$story_data           = $this->get_story_item_data( $story_id );
+		$single_story_classes = $this->get_single_story_classes();
+		$show_story_player    = true !== $this->attributes['show_story_poster'] && $this->is_view_type( 'grid' )
```
Comment (dhaval-parekh) [inline includes/Stories_Renderer/Renderer.php:275]: Use a round bracket around the condition. Please update here any everywhere to maintain consistancy.

File: includes/Stories_Renderer/Renderer.php [hunk 23]
```diff
@@ -0,0 +1,386 @@
...
+		$width        = ! empty( $story_data['width'] ) ? absint( $story_data['width'] ) : 360;
+		$poster_style = sprintf( 'background-image: url(%1$s);', esc_url_raw( $story_data['poster'] ) );
+		$poster_style = ( true === $this->is_view_type( 'carousel' ) ) ?
+			sprintf( '%1$s width: %2$spx; height: %3$spx', $poster_style, $width, $height ) : $poster_style;
+
+		$list_view_image_alignment = ! empty( $this->attributes['list_view_image_alignment'] ) ? sprintf( 'image-align-%1$s', $this->attributes['list_view_image_alignment'] ) : '';
+
+		?>
+		<a class="<?php echo( esc_attr( $list_view_image_alignment ) ); ?>"
+			href="<?php echo( esc_url_raw( $story_data['url'] ) ); ?>"
```
Comment (dhaval-parekh) [inline includes/Stories_Renderer/Renderer.php:309]: Use `esc_url()` for escaping URLs.  Please update is everywhere.

File: tests/phpunit/includes/Test_Renderer.php [hunk 24]
```diff
@@ -0,0 +1,43 @@
...
+ * Generic_Renderer class.
+ */
+class Test_Renderer extends Renderer {
+
+	/**
+	 * Render method
+	 *
+	 * @return void
+	 */
+	public function render(){}
```
Comment (dhaval-parekh) [inline tests/phpunit/includes/Test_Renderer.php:41]: Keep the consistency in code format.

```
public function render() {}
```

File: tests/phpunit/tests/Stories.php [hunk 25]
```diff
@@ -0,0 +1,133 @@
...
+		self::$testee->render();
+		ob_end_clean();
+
+		$this->assertInstanceOf( Generic_Renderer::class, self::$testee->renderer );
+	}
+
+	/**
+	 * Test that get_stories method returns valid story.
+	 */
+	public function test_get_stories_returns_valid_story() {
```
Comment (dhaval-parekh) [inline tests/phpunit/tests/Stories.php:112]: Please also mention which function this test case is covering.

```
@covers ::get_stories_returns_valid_story
```

File: tests/phpunit/tests/Stories_Renderer/Renderer.php [hunk 26]
```diff
@@ -0,0 +1,368 @@
...
+		);
+		$renderer = new Test_Renderer( $this->stories );
+
+		$output = $this->call_private_method( $renderer, 'get_view_type' );
+
+		$this->assertEquals( 'grid', $output );
+
+	}
+
+	public function test_get_story_item_data() {
```
Comment (dhaval-parekh) [inline tests/phpunit/tests/Stories_Renderer/Renderer.php:126]: Please add a function document.

**Review by dhaval-parekh — CHANGES_REQUESTED** [review]

**Review by dhaval-parekh — CHANGES_REQUESTED** [review]

**Comment by ankitrox** [comment]
Closing in favour of #152
